examples/nucleo_g474re/json/main.cpp

```diff
@@ -0,0 +1,186 @@
+
+		// JSON is fairly self-annotated, so we might not even need the original struct
+		MODM_LOG_INFO << "\nRetrieved from string JSON has fields ... ";
+		for (auto& [key, value] : retrieved_alice_json.items())
+		{
+			MODM_LOG_INFO << key.c_str() << ", ";
+		}
+		MODM_LOG_INFO << modm::endl;
+		MODM_LOG_INFO.flush();
+
+		// check if retrieved struct is same
+		auto retrieved_alice_struct = retrieved_alice_json.template get<data::person>();
+		if (retrieved_alice_struct == alice_struct)
+		{
+			MODM_LOG_INFO << "\nRetrieved from string and original struct are identical"
+						  << modm::endl;
+		}
+		MODM_LOG_INFO.flush();
+	}
+
+	// basic binary JSON usage
+	{
+		// convert to BSON
+		auto alice_binary = json::to_bson(alice_json);
+
+		MODM_LOG_INFO << "\nBinary JSON (BSON): ";
+		for (const auto& byte : alice_binary) { MODM_LOG_INFO.printf(" 0x%02x", byte); }
+		MODM_LOG_INFO << modm::endl;
+
+		// put BSON in first flash page after program
+		uint32_t err{0};
+		const size_t page_start = Flash::getPage(reinterpret_cast<uint32_t>(&__rom_end)) + 1;
+		const size_t num_bytes = alice_binary.size();
+		const size_t flash_page_size = Flash::getSize(page_start);
+		const size_t end_page = page_start + (num_bytes + flash_page_size - 1) / flash_page_size;
+		if (end_page >= max_flash_pages)
+		{
+			MODM_LOG_ERROR << "\nRequested flash end page exceeds flash [" << page_start << ", "
+						   << end_page << ")" << modm::endl;
+			MODM_LOG_INFO.flush();
+			while (1);
+		}
+
+		// erase the pages before programming
+		MODM_LOG_INFO << "\nErasing flash sectors [" << page_start << ", " << end_page << ")"
+					  << modm::endl;
+		MODM_LOG_INFO.flush();
+		if (not Flash::unlock())
+		{
+			MODM_LOG_INFO << "Flash unlock failed!" << modm::endl;
+			while (1);
+		}
+		for (size_t page{page_start}; page < end_page; page++) err |= Flash::erase(page);
+		if (err != 0)
+		{
+			MODM_LOG_ERROR << "\nThere was an error while erasing flash!" << modm::endl;
+			MODM_LOG_INFO.flush();
+			while (1);
+		}
+
+		// pad the data with zeros to fit flash words
+		size_t unpadded_size{alice_binary.size()};
+		size_t word_size{sizeof(Flash::MaxWordType)};
+		if (auto padding = (word_size - unpadded_size % word_size) % word_size; padding > 0)
+		{
+			alice_binary.resize(alice_binary.size() + padding, 0);
+		}
+
+		// now, write the padded data
+		MODM_LOG_INFO << "\nWriting, word size: " << word_size
+					  << ", num of bytes (payload): " << unpadded_size
+					  << ", num of bytes (padded): " << alice_binary.size() << "... ";
+
+		for (uint32_t src_addr = reinterpret_cast<uint32_t>(&alice_binary[0]),
+					  dst_addr{reinterpret_cast<uint32_t>(Flash::getAddr(page_start))};
+			 src_addr < (reinterpret_cast<uint32_t>(&alice_binary[0]) + alice_binary.size());
+			 src_addr += sizeof(Flash::MaxWordType), dst_addr += sizeof(Flash::MaxWordType))
+		{
+			err |= Flash::program(dst_addr, *(Flash::MaxWordType*)src_addr);
+		}
```

Comment:
That flash API is really messy to work with... It would really be nice to also have overloads that take pointers for memory locations instead of integers. Even `void*` would be a big type safety improvement over `uintptr_t` because we could get rid of all the `reinterpret_cast`s and mixing up address and data would be a hard compile error instead of a silent runtime one.

If we had a `memcpy` like overload of `Flash::program` the user wouldn't need to write that verbose loop with all the casts. I'll look into it tomorrow. Coincidentally, I had to write similar code at work today storing data in flash on a STM32F1 and wasn't very pleased with the API.

Also `*(Flash::MaxWordType*)src_addr` is [undefined behavior](https://en.cppreference.com/w/cpp/language/reinterpret_cast#Type_aliasing) in both C and C++ if `src_addr` doesn't point to an actual object of type `MaxWordType` and `MaxWordType` is neither of `char`, `unsigned char` or `std::byte`.

The usual conforming way to write this is using a memcpy that is optimized away by the compiler. Code generation should be [equivalent](https://gcc.godbolt.org/z/shdPvTK8d) with at least `-Og`. On Cortex-M0, for example, the original code will crash if the source data is not properly aligned, but this is always UB even if all data is sufficiently aligned.

Reply:
I think I blindly copied the flash access part from [one of the flash examples](https://github.com/search?q=repo%3Amodm-io%2Fmodm%20*(Flash%3A%3AMaxWordType*)src_addr&type=code) . 

>  Even void* would be a big type safety improvement over uintptr_t because we could get rid of all the reinterpret_casts and mixing up address and data would be a hard compile error instead of a silent runtime one.

Yes, happened to me quite a few times while I tried figuring out how the flash works.

Reply:
Ok, I've changed the loop to `memcpy`; is that what you had in mind?
This is still not fixing the problem with the `flash::program` api imho but definitely looks nicer than what I had before.

```
const auto flash_write_base_addr{reinterpret_cast<uint32_t>(Flash::getAddr(page_start))};
for (size_t ii = 0; ii < alice_binary.size(); ii += sizeof(Flash::MaxWordType))
{
	Flash::MaxWordType outdata;
	memcpy(&outdata, &alice_binary[ii], sizeof(Flash::MaxWordType));
	err |= Flash::program(flash_write_base_addr+ii, outdata);
}
```